folly/concurrency/ConcurrentHashMap.h

```diff
@@ -228,7 +228,7 @@ class ConcurrentHashMap {
       auto seg = segments_[i].load(std::memory_order_relaxed);
       if (seg) {
         seg->~SegmentT();
-        Allocator().deallocate((uint8_t*)seg, sizeof(SegmentT));
+        deallocateOverAligned<decltype(Allocator()), alignof(SegmentT)>(Allocator(), (typename std::allocator_traits<Allocator>::pointer)seg, sizeof(SegmentT));
```

Comment:
The following works with `SegAlloc` defined as above. 
`SegmentT* newseg = allocateOverAligned(SegAlloc(), 1);`
`deallocateOverAligned(SegAlloc(), newseg, 1);`

I see that rebind_alloc creates a std::allocator<SegmentT> instead of the default type std::allocator<uint8_t>. Does the std::allocator with the SegmentT template parameter ensure that the when a newseg is allocated and de allocated as above, it is aligned correctly ? . Also, the 1 in the parameters, does that mean size of one SegmentT, because I see that the above statements work with the parameter as 1 and sizeof(SegmentT).

Reply:
Good support for over-aligned types only really arrived in C++17. I don't know if folly has adopted it yet, but it was not available when I left FB two years ago.

When I wrote `allocateOverAligned`, allocators used by folly were not required to deal with over-aligned types, and std::allocator did not. C++17 added an alignment parameter to `::operator new`, which now makes it possible to do the right thing (but not if you use a byte allocator!). `allocateOverAligned` chooses from a variety of strategies at compile time so that it is zero-overhead when possible and always honors alignment. It looks like it doesn't know about the two-arg `::operator new`; it would be good to add that too.

In this case kAlign (64) > kBaseAlign (16), so it will call `aligned_malloc`, which calls `posix_memalign` (implemented by jemalloc).

Reply:
BTW, `operator new` is just a fancy malloc, it is not the same thing as saying `new T`

Reply:
sounds good, thanks, are the failing checks land blocking.
Will land the diff onto master, pull the changes, checkout to a branch `rockset_v2021.02.01.02`, add a tag `rockset_v2021.02.01.07` and then use that in the rs code.